### ISSUE TITLE
Unable to login using oAuth2/OKTA with docker-compose prod package - Error 'User could not be found'

### ISSUE THREAD

**Issue by maznag** [jhipster/generator-jhipster#7612]
##### **Overview of the issue**

Login doesn't work with oAuth2/Okta using docker-compose: Error 500 - 'User could not be found'. But, login works well when using keycloak instead of OKTA. The error occurred only with prod profile and docker-compose. 

##### **Motivation for or Use Case**

Testing OIDC

##### **Reproduce the error**

- Generate o Monolith app with oAuth2 option
- Change OpenID Connect Settings to use OKTA instead of Keycloak: src/main/docker/app.yml and src/main/resources/application.yml
- build docker image using './mvnw verify -Pprod dockerfile:build -DskipTests'
- Start app using 'docker-compose -f src/main/docker/app.yml up -d' 

##### **Related issues**

#7609

##### **JHipster Version(s)**

```
jhipster@5.0.0-beta.1

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.test.monolith",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "5.0.0-beta.1",
    "baseName": "monolith",
    "packageName": "com.test.monolith",
    "packageFolder": "com/test/monolith",
    "serverPort": "8080",
    "authenticationType": "oauth2",
    "cacheProvider": "hazelcast",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "clientFramework": "angularX",
    "useSass": true,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en",
      "fr"
    ]
  }
}
</pre>
</details>

##### **Environment and Tools**

java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)

git version 2.13.1
node: v8.11.1
npm: 5.6.0
yeoman: 2.0.0
yarn: 1.6.0
Docker version 18.03.0-ce, build 0520e24
docker-compose version 1.20.1, build 5d8c71b

##### **Browsers and Operating System**

- MacOS 10.13.4
- Chrome

**Comment by jdubois** [comment]
And did you create the user, as documented on https://www.jhipster.tech/security/#oauth2 ?

**Comment by jdubois** [comment]
Thanks @maznag - this has been heavily refactored by @mraible but as he works for Okta I'm pretty sure he made sure this was working! @mraible can I assign the issue to you?

**Comment by mraible** [comment]
Yes, please assign it to me. I’ll try to make time to look at it next week.

**Comment by maznag** [comment]
Thanks @jdubois, @mraible

**Comment by mraible** [comment]
@maznag I tried to reproduce this issue tonight with the latest code from the master branch and I was unable to. It all works with my Okta configuration. Did you complete the following steps to configure your group settings for Okta?

> Create a `ROLE_ADMIN` and `ROLE_USER` group (**Users** > **Groups** > **Add Group**) and add users to them. You can use the account you signed up with, or create a new user (**Users** > **Add Person**). Navigate to **API** > **Authorization Servers**, click the **Authorization Servers** tab and edit the default one. Click the **Claims** tab and **Add Claim**. Name it “groups” or “roles”, and include it in the ID Token. Set the value type to “Groups” and set the filter to be a Regex of `.*`.

From https://www.jhipster.tech/security/

**Comment by maznag** [comment]
@mraible Yes everything is configured correctly and Okta works well with JHipster in dev profile but I have the problem only when I try to start the generated application with docker-compose. I also tested with the latest code from the master branch.

The login works well when I use ./mvnw to start the generated application but I have the error "User could not be found' when I use 'docker-compose -f src/main/docker/app up -d'

**Comment by mraible** [comment]
I used Docker compose and an existing user in my Okta tenant. Are you using a person that’s not in the ROLE_USER or ROLE_ADMIN group?

I need to be able to reproduce the problem in order to fix it! ;)

**Comment by maznag** [comment]
I Use my personnel account in My Okta tenant that's in the ROLE_USER and ROLE_ADMIN groups.
<img width="821" alt="screen shot 2018-05-21 at 15 11 17" src="https://user-images.githubusercontent.com/23117630/40309564-3dc63bf6-5d0a-11e8-8d54-4eb1b956bdfb.png">

**Comment by mraible** [comment]
Do you experience the same issue if you run with the prod profile outside of Docker? Here's the `.yo-rc.json` I used to create my app.

```json
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.okta.developer",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "5.0.0-beta.1",
    "baseName": "odoc",
    "packageName": "com.okta.developer",
    "packageFolder": "com/okta/developer",
    "serverPort": "8080",
    "authenticationType": "oauth2",
    "cacheProvider": "ehcache",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "clientFramework": "angularX",
    "useSass": false,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ]
  }
}
```

**Comment by mraible** [comment]
If it doesn’t work in docker-compose, it might be caused by the variables set in src/main/docker/app.yml. If you already have them defined in src/main/resources/config/application.yml, there’s no need to override them in app.yml.

I’d try deleting the “SECURITY_OAUTH2_*” keys in app.yml and see if that fixes it.

**Comment by maznag** [comment]
Awesome ! thanks very much @mraible. It works well now with docker-compose after deleting “SECURITY_OAUTH2_*” keys in app.yml or in existing docker-compose.yml. But, why did it work well before without having to delete these keys ?

It may be necessary to modify the generator so as not to duplicate these keys in the docker-compose.yml file. Do you agree ? Or can I close this issue since it's resolved by deleting “SECURITY_OAUTH2_*” keys ?

thanks again !

**Comment by mraible** [comment]
In the situation you describe, there is no `docker-compose.yml`. Did you generate one using the sub-generator?

I believe the reason I put the keys in `app.yml` (and likely `docker-compose.yml`) is so you can override them with environment variables. When I demo Okta in JHipster, I usually don't edit `application.yml`. Instead, I set a bunch of environment variables to point to my Okta tenant. This has the benefit of not storing my Okta tenant information in GitHub (as it will still have the default Keycloak settings).

Here's a sample `okta.env` that I `source` before starting my JHipster apps:

```
#!/bin/bash

# Okta with JHipster
export SECURITY_OAUTH2_CLIENT_ACCESS_TOKEN_URI="https://dev-158606.oktapreview.com/oauth2/default/v1/token"
export SECURITY_OAUTH2_CLIENT_USER_AUTHORIZATION_URI="https://dev-158606.oktapreview.com/oauth2/default/v1/authorize"
export SECURITY_OAUTH2_RESOURCE_USER_INFO_URI="https://dev-158606.oktapreview.com/oauth2/default/v1/userinfo"
export SECURITY_OAUTH2_RESOURCE_TOKEN_INFO_URI="https://dev-158606.oktapreview.com/oauth2/default/v1/introspect"
export SECURITY_OAUTH2_RESOURCE_JWK_KEY_SET_URI="https://dev-158606.oktapreview.com/oauth2/default/v1/keys"
export SECURITY_OAUTH2_CLIENT_CLIENT_ID="XXX"
export SECURITY_OAUTH2_CLIENT_CLIENT_SECRET="YYY"
```

**Comment by maznag** [comment]
Yes, In this situation I just forgot to mention `docker-compose.yml`. But this issue is occurred for both:
- `docker-compose -f src/main/docker/app.yml up`
- Or when I generate `dokcer-compose.yml` with sub-generator `jhipster docker-compose`

In my case I put directly my okta settings in the `application.yml` and `app.yml` files. But, you're right it's better to use environment variables to point to Okta tenant. thanks again for this tip.

I close then this issue since it's resolved by deleting `SECURITY_OAUTH2_*` keys in `app.yml` or using environment variables to point to Okta tenant.

**Comment by mraible** [comment]
I'm getting this issue now too, so re-opening.
